### PR TITLE
Version 1.1.0

### PR DESCRIPTION
**Pull request by ssestak**
- Template Updates
   - Removed SwiftLint warnings
   - Omitted application prefix in naming
   - Dropped class prefix in mock view model events
- Added commentary explaining anti-pattern of adding view modifiers in coordinator root view
- Alerts no longer managed by coordinators; their presentation is now individual view's responsibility
- Scene Delegate
   - Introduced view modifier in helpers
- Destination IDs
   - Integrated `@EnumIdentifiersGenerator` macro for auto-generating code for `Hashable` and `Identifiable` protocol conformance
   - The macro produces nested enum for ID generation for enum cases
- Introduced FlowProvider in helpers
- For coordinators, integrated `ModalCoverModel<Destination>` to avoid UI confusion regarding sheet and full screen cover presentation order

Prerequisites for Merging:
- New projects must specify a key-value pair for `LaunchScreen` in `Info.plist` to ensure proper display
- Macro 
   - Requires importing `EnumIdentifiersGenerator` currently, but ideally, it should be part of `FuturedArchitecture`
      -  _UPDATE_: Macro moved into a separate [git repository](https://github.com/futuredapp/futured-macros)
   - It doesn't account for situation when Destination ID needs to be computed using an associated value
      - _UPDATE_: DONE

_Macro Expansion Example:_
image

### CODE REVIEW

**Review by samoilyk — DISMISSED** [review]

**Review by samoilyk — DISMISSED** [review]

File: Sources/FuturedHelpers/Helpers/CoordinatorSceneFlowProvider.swift [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+
+/** A protocol to represent navigable destinations in a reusable scene flow
+ ```
+ protocol TemplateFlowDestination: CoordinatorSceneFlowDestination {
+    static var destination: Self { get }
+    static var end: Self { get }
+ }
+ ```
+ **/
+protocol CoordinatorSceneFlowDestination: Hashable, Identifiable, Equatable {}
```
Comment (mikolasstuchlik) [inline Sources/FuturedHelpers/Helpers/CoordinatorSceneFlowProvider.swift]: Can be 
```suggestion
typealias CoordinatorSceneFlowDestination = Hashable & Identifiable & Equatable
```

File: Sources/FuturedHelpers/Helpers/CoordinatorSceneFlowProvider.swift [hunk 2]
```diff
@@ -0,0 +1,210 @@
+//
+//  CoordinatorSceneFlowProvider.swift
+//
+//  Created by Simon Sestak on 31/07/2024.
+//
+
+import SwiftUI
+
+/** A protocol to represent navigable destinations in a reusable scene flow
```
Comment (mikolasstuchlik) [inline Sources/FuturedHelpers/Helpers/CoordinatorSceneFlowProvider.swift]: Please provide a documentation

File: Sources/FuturedHelpers/Helpers/CoordinatorSceneFlowProvider.swift [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+ ```
+ protocol TemplateFlowDestination: CoordinatorSceneFlowDestination {
+    static var destination: Self { get }
+    static var end: Self { get }
+ }
+ ```
+ **/
+protocol CoordinatorSceneFlowDestination: Hashable, Identifiable, Equatable {}
+
+/**
```
Comment (mikolasstuchlik) [inline Sources/FuturedHelpers/Helpers/CoordinatorSceneFlowProvider.swift:10]: Please provide a shorter documentation. 
In my opinion, code examples should be no more than 10 lines of code in once block.

Having a template would be great, as discussed on the call.

File: Sources/FuturedHelpers/Helpers/CoordinatorSceneFlowProvider.swift [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+    func scene(for destination: Destination) -> DestinationViews
+
+    var onNavigateToDestination: (Destination) -> Void { get }
+    var onPop: () -> Void { get }
+
+    var onPresentSheet: ((Destination) -> Void)? { get }
+    var onDismissSheet: (() -> Void)? { get }
+    var onPresentFullscreenCover: ((Destination) -> Void)? { get }
+    var onDismissFullscreenCover: (() -> Void)? { get }
+    var onPopToDestination: ((Destination?) -> Void)? { get }
```
Comment (mikolasstuchlik) [inline Sources/FuturedHelpers/Helpers/CoordinatorSceneFlowProvider.swift]: Please document, what is the significance of `nil` argument.

File: Sources/FuturedHelpers/Helpers/CoordinatorSceneFlowProvider.swift [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+    private lazy var templateSceneProvider: TemplateSceneFlowProvider = {
+        TemplateSceneFlowProvider(container: container, onNavigateToDestination: { [weak self] destination in
+            if destination == .end {
+                self?.navigate(to: .flowSpecificDestinationAfterEmbededFlow)
+            } else {
+                self?.navigate(to: destination)
+            }
+        }, onPop: { [weak self] in
+            self?.pop()
+        })
```
Comment (mikolasstuchlik) [inline Sources/FuturedHelpers/Helpers/CoordinatorSceneFlowProvider.swift]: This is a matter of opinion, but I prefer following multi-closure function calls:
```suggestion
        TemplateSceneFlowProvider(
            container: container, 
            onNavigateToDestination: { [weak self] destination in
                if destination == .end {
                    self?.navigate(to: .flowSpecificDestinationAfterEmbededFlow)
                } else {
                    self?.navigate(to: destination)
                }
            },
            onPop: { [weak self] in
                self?.pop()
            }
        )
```

File: Sources/FuturedHelpers/Helpers/CoordinatorSceneFlowProvider.swift [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+        onDismissFullscreenCover?()
+    }
+
+    func pop(to destination: Destination?) {
+        onPopToDestination?(destination)
+    }
+
+    func show(error: Error) {
+        onShowError?(error)
+    }
```
Comment (mikolasstuchlik) [inline Sources/FuturedHelpers/Helpers/CoordinatorSceneFlowProvider.swift]: I don't understand the reason for this extension. If the _only_ reason for this extension existence is to have _nicer_ API, I would suggest to rename the closures.

I think that reasoning the code is made _more_ difficult, not less, and thus I would only keep the closures.

For sake of transparency, I have also renamed a variable in similar manner, but in my case, I have introduced different access modifiers to separate contexts and I've only done it for 1 variable.

File: Templates/SwiftUI App.xctemplate/ExampleFlowCoordinator.swift [hunk 7]
```diff
@@ -40,11 +38,8 @@ final class ExampleFlowCoordinator: NavigationStackCoordinator {
 }
 
 extension ExampleFlowCoordinator {
+    @EnumIdentable
     enum Destination: String, Hashable, Identifiable {
```
Comment (mikolasstuchlik) [inline Templates/SwiftUI App.xctemplate/ExampleFlowCoordinator.swift]:
```suggestion
    enum Destination: Hashable, Identifiable {
```

We also might make this macro add the conformances, as previously discussed. 

https://github.com/futuredapp/futured-macros/pull/1

File: Templates/Tab Coordinator.xctemplate/___VARIABLE_tabCoordinatorIdentifier___TabCoordinator.swift [hunk 8]
```diff
@@ -52,3 +44,10 @@ final class ___VARIABLE_tabCoordinatorIdentifier___TabCoordinator: TabCoordinato
         }
     }
 }
+
+extension ___VARIABLE_tabCoordinatorIdentifier___TabCoordinator {
+    @EnumIdentable
+    enum Destination: String, Hashable, Identifiable {
+        case destination
+    }
```
Comment (mikolasstuchlik) [inline Templates/Tab Coordinator.xctemplate/___VARIABLE_tabCoordinatorIdentifier___TabCoordinator.swift]:
```suggestion
    @EnumIdentable
    enum Destination: Hashable, Identifiable {
        case destination
    }
```

**Review by mikolasstuchlik — CHANGES_REQUESTED** [review]
Overall a great piece of pioneering work!

File: Sources/FuturedHelpers/Helpers/CoordinatorSceneFlowProvider.swift (hunk 6 above)
Comment (mikolasstuchlik) [reply Sources/FuturedHelpers/Helpers/CoordinatorSceneFlowProvider.swift]: On second thought, I almost lean to a conclusion, that `protocol` is just a bad tool for this job. I would almost suggest to use `class` inheritance.

**Review by samoilyk — DISMISSED** [review]

File: Sources/FuturedHelpers/Helpers/CoordinatorSceneFlowProvider.swift [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+
+/**
+ A typealias representing navigable destinations in a reusable scene flow.
+
+ # Notes: #
+ 1. If the scenes defined in the provider can continue with other scenes, define an end destination.
+    - This end destination should trigger a scene display outside of the scene provider.
+ 2. Other destinations will be used to display scenes defined within the scene provider.
+
+ # Example #
```
Comment (mikolasstuchlik) [inline Sources/FuturedHelpers/Helpers/CoordinatorSceneFlowProvider.swift]:
```suggestion
 # Example #
  Declare a `protocol` conforming to the composition. The protocol should only be conformed to by enums. We use the requirements to force appearance of certain cases: [SE-0280](https://github.com/swiftlang/swift-evolution/blob/main/proposals/0280-enum-cases-as-protocol-witnesses.md).
```

File: Sources/FuturedHelpers/Helpers/CoordinatorSceneFlowProvider.swift [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+ ```
+ protocol TemplateFlowDestination: CoordinatorSceneFlowDestination {
+    static var destination: Self { get }
+    static var end: Self { get }
```
Comment (mikolasstuchlik) [inline Sources/FuturedHelpers/Helpers/CoordinatorSceneFlowProvider.swift]: We probably should use naming which does not appear to have a special meaning, if it is only an example.

```suggestion
    static var someDestination: Self { get }
    static var otherDestionation: Self { get }
```

File: Sources/FuturedHelpers/Helpers/CoordinatorSceneFlowProvider.swift [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+ }
+ ```
+ */
+public typealias CoordinatorSceneFlowDestination = Hashable & Identifiable & Equatable
+
+
+/**
+ A protocol providing an interface for reusable scene flow providers.
```
Comment (mikolasstuchlik) [inline Sources/FuturedHelpers/Helpers/CoordinatorSceneFlowProvider.swift]: What is a _reusable scene flow provider_? Is my suggestion correct?

```suggestion
 A protocol providing an interface for reusable scene flow providers. *Reusable scene flow provider* is part of a scene flow, which can be used as a part of more *Flow Coordinators*. The shared section of the flow is taken out of the *Flow Coordinator* and placed into a class conforming to `CoordinatorSceneFlowProvider`. 
```

File: Templates/Flow Coordinator.xctemplate/___VARIABLE_flowCoordinatorIdentifier___FlowCoordinator.swift [hunk 12]
```diff
@@ -30,11 +30,8 @@ final class ___VARIABLE_flowCoordinatorIdentifier___FlowCoordinator: NavigationS
 }
 
 extension ___VARIABLE_flowCoordinatorIdentifier___FlowCoordinator {
-    enum Destination: String, Hashable, Identifiable {
+    @EnumIdentable
+    enum Destination: Hashable, Identifiable {
```
Comment (mikolasstuchlik) [inline Templates/Flow Coordinator.xctemplate/___VARIABLE_flowCoordinatorIdentifier___FlowCoordinator.swift:34]: I still think we should make the `@EnumIdentable` add the conformance. I don't think it's MUST-HAVE at this point, but please, file an Issue to this effect.

File: Templates/SceneFlowProvider.xctemplate/___VARIABLE_sceneFlowProviderIdentifier___SceneFlowProvider.swift [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+import FuturedArchitecture
+import FuturedHelpers
+import SwiftUI
+
+protocol ___VARIABLE_sceneFlowProviderIdentifier___FlowDestination: CoordinatorSceneFlowDestination {
+    static var destination: Self { get }
+    static var end: Self { get }
+}
+
+
```
Comment (mikolasstuchlik) [inline Templates/SceneFlowProvider.xctemplate/___VARIABLE_sceneFlowProviderIdentifier___SceneFlowProvider.swift]:
```suggestion

```

File: Templates/SceneFlowProvider.xctemplate/___VARIABLE_sceneFlowProviderIdentifier___SceneFlowProvider.swift [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+    }
+}
+
+extension ___VARIABLE_sceneFlowProviderIdentifier___SceneFlowProvider {
+    @EnumIdentable
+    enum Destination: ___VARIABLE_sceneFlowProviderIdentifier___FlowDestination {
+        case destination
+        case end
+    }
+}
```
Comment (mikolasstuchlik) [inline Templates/SceneFlowProvider.xctemplate/___VARIABLE_sceneFlowProviderIdentifier___SceneFlowProvider.swift]: I don't see why we need to do this. The `Destination` should be only a generic parameter of the `___VARIABLE_sceneFlowProviderIdentifier___SceneFlowProvider` and conformed to by the Destinations of the _Flow cooridnators_.

If I am not correct, why do we need `___VARIABLE_sceneFlowProviderIdentifier___FlowDestination` at all?

**Review by mikolasstuchlik — CHANGES_REQUESTED** [review]

**Comment by mikolasstuchlik** [comment]
It is still not fully clear to me, how the Scene Provider should work. I don't want to hinder the acceptance because of the documentation. I have made some requests and the rest of it can be deferred to an Issue (on both FuturedKit and macros).

However, I'm reluctant to merge Scene Provider in this state.

**Review by mikolasstuchlik — APPROVED** [review]

**Comment by mikolasstuchlik** [comment]
86580cded392542b51fe48041237db67

**Pull request merged**